prov/opx/src/fi_opx_ep.c

```diff
@@ -0,0 +1,2210 @@
+	.tx_size_left   = fi_no_tx_size_left
+};
+
+int fi_opx_alloc_default_rx_attr(struct fi_rx_attr **rx_attr)
+{
+	struct fi_rx_attr *attr;
+
+	attr = calloc(1, sizeof(*attr));
+	if (!attr)
+		goto err;
+
+	attr->caps 	= FI_OPX_DEFAULT_RX_CAPS;
+	attr->mode 	= FI_CONTEXT2 | FI_ASYNC_IOV;
+	attr->op_flags 	= 0;
+	attr->msg_order = FI_OPX_DEFAULT_MSG_ORDER;
+	attr->comp_order = FI_ORDER_NONE;
+	attr->total_buffered_recv = FI_OPX_HFI1_PACKET_MTU;
+	attr->size 	= SIZE_MAX; //FI_OPX_RX_SIZE;
+	attr->iov_limit = FI_OPX_IOV_LIMIT;
+
+	*rx_attr = attr;
+
+	return 0;
+err:
+	errno = FI_EINVAL;
+	return -errno;
+}
+
+int fi_opx_check_rx_attr(struct fi_rx_attr *attr)
+{
+	/* TODO: more error checking of rx_attr */
+	if (attr->comp_order && attr->comp_order == FI_ORDER_STRICT) {
+		FI_WARN(fi_opx_global.prov, FI_LOG_EP_DATA,
+				"unavailable [bad rx comp_order (%lx)] ",
+				attr->comp_order);
+		goto err;
+	}
+
+	return 0;
+err:
+	errno = FI_EINVAL;
+	return -errno;
+}
+
+int fi_opx_alloc_default_tx_attr(struct fi_tx_attr **tx_attr)
+{
+	struct fi_tx_attr *attr;
+
+	attr = calloc(1, sizeof(*attr));
+	if (!attr)
+		goto err;
+
+	attr->caps	= FI_OPX_DEFAULT_TX_CAPS;
+	attr->mode	= FI_CONTEXT2 | FI_ASYNC_IOV;
+	attr->op_flags	= FI_TRANSMIT_COMPLETE;
+	attr->msg_order	= FI_OPX_DEFAULT_MSG_ORDER;
+	attr->comp_order = FI_ORDER_NONE;
+	attr->inject_size = FI_OPX_HFI1_PACKET_IMM;
+	attr->size	= SIZE_MAX;
+	attr->iov_limit = FI_OPX_IOV_LIMIT;
+	attr->rma_iov_limit = 1;
+
+	*tx_attr = attr;
+
+	return 0;
+err:
+	errno = FI_EINVAL;
+	return -errno;
+}
+
+int fi_opx_check_tx_attr(struct fi_tx_attr *attr)
+{
+	if (attr->inject_size > FI_OPX_HFI1_PACKET_IMM) {
+		FI_LOG(fi_opx_global.prov, FI_LOG_DEBUG, FI_LOG_EP_DATA,
+				"unavailable [bad inject_size (%lu)]",
+				attr->inject_size);
+		goto err;
+	}
+	/* TODO: more error checking of tx_attr */
+
+	if (attr->comp_order && attr->comp_order == FI_ORDER_STRICT) {
+		FI_LOG(fi_opx_global.prov, FI_LOG_DEBUG, FI_LOG_EP_DATA,
+				"unavailable [bad tx comp_order (%lx)] ",
+				attr->comp_order);
+		goto err;
+       }
+
+	return 0;
+err:
+	errno = FI_EINVAL;
+	return -errno;
+}
+
+int fi_opx_alloc_default_ep_attr(struct fi_ep_attr **ep_attr)
+{
+	struct fi_ep_attr *attr;
+
+	attr = calloc(1, sizeof(*attr));
+	if (!attr)
+		goto err;
+
+	attr->type		= FI_EP_RDM;
+	attr->protocol		= FI_PROTO_OPX;
+	attr->protocol_version	= FI_OPX_PROTOCOL_VERSION;
+	attr->max_msg_size	= FI_OPX_MAX_MSG_SIZE;
+	attr->msg_prefix_size 	= 0;
+	attr->max_order_raw_size= FI_OPX_MAX_ORDER_RAW_SIZE;
+	attr->max_order_war_size= FI_OPX_MAX_ORDER_WAR_SIZE;
+	attr->max_order_waw_size= FI_OPX_MAX_ORDER_WAW_SIZE;
+	attr->mem_tag_format 	= FI_OPX_MEM_TAG_FORMAT;
+	attr->tx_ctx_cnt	= 1;
+	attr->rx_ctx_cnt	= 1;
+
+	*ep_attr = attr;
+
+	return 0;
+err:
+	errno = FI_EINVAL;
+	return -errno;
+}
+
+int fi_opx_check_ep_attr(struct fi_ep_attr *attr)
+{
+	switch(attr->protocol) {
+		case FI_PROTO_UNSPEC:
+			attr->type  	= FI_EP_RDM;
+			attr->protocol	= FI_PROTO_OPX;
```

Comment:
The input `attr` is modified here. In `fi_opx_getinfo()`, the funciton `fi_opx_check_info()` is called with `hints` as the parameter, and thus hints is modified and the protocol setting will fail all other providers.

Reply:
This affects the cases that non-NULL hints is passed to the OPX provider. The `fi_info -f -p` command line falls into this category. `fi_info` w/o argument works fine because hints is NULL. `fi_info -p verbs` works fine because the OPX provider is filtered out.

Reply:
@j-xiong 
> This affects the cases that non-NULL hints is passed to the OPX provider. The `fi_info -f -p` command line falls into this category.

!@#$%!!! Yup. That's the problem. Thanks for finding that!

Reply:
Was the const cast away through this path?

Reply:
There's no specific cast to remove the const from `fi_opx_check_info(const struct fi_info *info)` but we don't actually change the values in *info, we change the values pointed to by *(info->ep_attr) which isn't protected by the const on the parent pointer.